pkg/server/plugin/keymanager/gcpkms/gcpkms.go

```diff
@@ -0,0 +1,1040 @@
+		Name:   keyEntry.cryptoKeyVersionName,
+		Digest: digest,
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to sign: %v", err)
+	}
+
+	// Perform integrity verification.
+	if int64(crc32Checksum(signResp.Signature)) != signResp.SignatureCrc32C.Value {
+		return nil, status.Error(codes.Internal, "error signing: response corrupted in-transit")
+	}
+
+	return &keymanagerv1.SignDataResponse{
+		Signature:      signResp.Signature,
+		KeyFingerprint: keyEntry.publicKey.Fingerprint,
+	}, nil
+}
+
+// createKey creates a new CryptoKey with a new CryptoKeyVersion.
+// If there is a specified IAM policy through the KeyPolicyFile configuration,
+// that policy is set to the created CryptoKey. If there is no IAM policy specified,
+// a default policy is constructed and attached. This function requests Cloud KMS
+// to get the public key of the created CryptoKeyVersion. A keyEntry is returned
+// with the CryptoKey, CryptoKeyVersion and public key.
+func (p *Plugin) createKey(ctx context.Context, spireKeyID string, keyType keymanagerv1.KeyType) (*keyEntry, error) {
+	p.entriesMtx.Lock()
+	defer p.entriesMtx.Unlock()
+
+	algorithm, err := cryptoKeyVersionAlgorithmFromKeyType(keyType)
+	if err != nil {
+		return nil, err
+	}
+
+	if entry, ok := p.entries[spireKeyID]; ok {
+		// Check if the algorithm has changed and update if needed
+		if entry.cryptoKey.VersionTemplate.Algorithm != algorithm {
+			entry.cryptoKey.VersionTemplate.Algorithm = algorithm
+			_, err := p.kmsClient.UpdateCryptoKey(ctx, &kmspb.UpdateCryptoKeyRequest{
+				CryptoKey: entry.cryptoKey,
+			})
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "failed to update CryptoKey with updated algorithm: %v", err)
+			}
+		}
+		cryptoKeyVersion, err := p.kmsClient.CreateCryptoKeyVersion(ctx, &kmspb.CreateCryptoKeyVersionRequest{
+			Parent: entry.cryptoKey.Name,
+			CryptoKeyVersion: &kmspb.CryptoKeyVersion{
+				State: kmspb.CryptoKeyVersion_ENABLED,
+			},
+		})
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to create CryptoKeyVersion: %v", err)
+		}
+		p.log.Debug("CryptoKeyVersion added", cryptoKeyNameTag, entry.cryptoKey.Name, cryptoKeyVersionNameTag, cryptoKeyVersion.Name)
+
+		pubKey, err := getPublicKeyFromCryptoKeyVersion(ctx, p.kmsClient, cryptoKeyVersion.Name)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to get public key: %v", err)
+		}
+
+		newKeyEntry := keyEntry{
+			cryptoKey:            entry.cryptoKey,
+			cryptoKeyVersionName: cryptoKeyVersion.Name,
+			publicKey: &keymanagerv1.PublicKey{
+				Id:          spireKeyID,
+				Type:        keyType,
+				PkixData:    pubKey,
+				Fingerprint: makeFingerprint(pubKey),
+			},
+		}
+
+		p.entries[spireKeyID] = newKeyEntry
+
+		if err := p.enqueueDestruction(entry.cryptoKeyVersionName); err != nil {
+			p.log.Error("Failed to enqueue CryptoKeyVersion for destruction", reasonTag, err)
+		}
+
+		return &newKeyEntry, nil
```

Comment:
Since we're effectively branching all of `createKey` on whether we have an existing crypto key or not, it might be easier to read if it were broken out into two further functions